assembly/src/assembler/tests.rs

```diff
@@ -124,90 +127,88 @@ fn nested_blocks() {
-        expected_mast_forest.ensure_node(r#false_node)
+        expected_mast_forest.add_node(r#false_node)
     };
     let r#if1 = {
         let r#if_node = MastNode::new_split(r#true1, r#false1, &expected_mast_forest);
-        expected_mast_forest.ensure_node(r#if_node)
+        expected_mast_forest.add_node(r#if_node)
     };
 
     let r#true3 = {
         let r#true_node = MastNode::new_basic_block(vec![Operation::Push(7u32.into())]);
-        expected_mast_forest.ensure_node(r#true_node)
+        expected_mast_forest.add_node(r#true_node)
     };
     let r#false3 = {
         let r#false_node = MastNode::new_basic_block(vec![Operation::Push(11u32.into())]);
-        expected_mast_forest.ensure_node(r#false_node)
+        expected_mast_forest.add_node(r#false_node)
     };
     let r#true2 = {
         let r#if_node = MastNode::new_split(r#true3, r#false3, &expected_mast_forest);
-        expected_mast_forest.ensure_node(r#if_node)
+        expected_mast_forest.add_node(r#if_node)
     };
 
     let r#while = {
         let push_basic_block_id = {
             let push_basic_block = MastNode::new_basic_block(vec![Operation::Push(23u32.into())]);
-            expected_mast_forest.ensure_node(push_basic_block)
+            expected_mast_forest.add_node(push_basic_block)
         };
         let body_node_id = {
             let body_node =
                 MastNode::new_join(exec_bar_node_id, push_basic_block_id, &expected_mast_forest);
 
-            expected_mast_forest.ensure_node(body_node)
+            expected_mast_forest.add_node(body_node)
         };
 
         let loop_node = MastNode::new_loop(body_node_id, &expected_mast_forest);
-        expected_mast_forest.ensure_node(loop_node)
+        expected_mast_forest.add_node(loop_node)
     };
     let push_13_basic_block_id = {
         let node = MastNode::new_basic_block(vec![Operation::Push(13u32.into())]);
-        expected_mast_forest.ensure_node(node)
+        expected_mast_forest.add_node(node)
     };
 
     let r#false2 = {
         let node = MastNode::new_join(push_13_basic_block_id, r#while, &expected_mast_forest);
-        expected_mast_forest.ensure_node(node)
+        expected_mast_forest.add_node(node)
     };
     let nested = {
         let node = MastNode::new_split(r#true2, r#false2, &expected_mast_forest);
-        expected_mast_forest.ensure_node(node)
+        expected_mast_forest.add_node(node)
     };
 
     let combined_node_id = combine_mast_node_ids(
         vec![before, r#if1, nested, exec_foo_bar_baz_node_id, syscall_foo_node_id],
         &mut expected_mast_forest,
     );
-    expected_mast_forest.set_entrypoint(combined_node_id);
-
-    let combined_node = &expected_mast_forest[combined_node_id];
+    let expected_program = Program::new(expected_mast_forest.into(), combined_node_id);
 
-    assert_eq!(combined_node.digest(), program.entrypoint_digest().unwrap());
+    assert_eq!(expected_program.hash(), program.hash());
```

Comment:
Should we not also check that the underlying `MastForest` has the correct number of `roots`? I think it should be 3: two for named procedures and one for the main procedure.

Also, to incorporate into the test a duplicate node and make sure it is handled correctly. For example, we could have 2 identical statement blocks and we should make sure we generate only a single node in the `MastForest` for both of them.

Reply:
> Should we not also check that the underlying MastForest has the correct number of roots? I think it should be 3: two for named procedures and one for the main procedure.

Done (although it is 5, due to `KERNEL` and `PROCEDURE` at the beginning)

Reply:
> Also, to incorporate into the test a duplicate node and make sure it is handled correctly. For example, we could have 2 identical statement blocks and we should make sure we generate only a single node in the MastForest for both of them.

Added test `duplicate_procedure()`